src/DynamicData.Tests/Utilities/ObservableSpy.cs

```diff
@@ -0,0 +1,183 @@
+﻿using System;
+using System.Diagnostics;
+using System.Linq;
+using System.Reactive.Disposables;
+using System.Reactive.Linq;
+using System.Threading;
+
+namespace DynamicData.Tests.Utilities;
+
+internal static class ObservableSpy
+{
+    private static readonly string ChangeSetEntrySpacing = Environment.NewLine + "\t";
+
+    /// <summary>
+    /// Spys on the given IObservable{T} by emitting logging information that is tagged with the current ThreadId for all related
+    /// events including every invocation on the Observer, subscriptions, disposes, and exceptions.
+    /// </summary>
+    /// <typeparam name="T">The type of the Observable.</typeparam>
+    /// <param name="source">The source IObservable to be Spied on.</param>
+    /// <param name="logger">The logger instance to use for logging.</param>
+    /// <param name="infoText">Optional text to include with each log message.</param>
+    /// <param name="formatter">Optional Value transformer to control how the observed values are logged.</param>
+    /// <param name="showSubs">Indicates whether or not subscription related messages should be emitted.</param>
+    /// <param name="showTimestamps">Indicates whether or not timestamps should be prepended to messages.</param>
+    /// <returns>An IObservable{T} with the Spy events included.</returns>
+    /// <remarks>Adapted from https://stackoverflow.com/q/20220755/.</remarks>
+    public static IObservable<T> Spy<T>(this IObservable<T> source, string? infoText = null, Action<string>? logger = null,
```

Comment:
This is super cool

Reply:
It is amazing. I wish I could take credit, but it's not my idea... I just fleshed it out a bit.  It is game-changing in terms of debugging Rx.  I did add the ChangeSet-specific overloads that allow it to peer into the individual changes though.  This has saved me a ton of time and hassle and helps get you right to the root of an issue.  I really needed its help for a couple of the test-cases and so I ported it over from my personal library.

I think you should consider promoting it to the main library instead of the test library, but I didn't want to be presumptuous.  Also, I don't think the interface is quite ready for mainstream.  It's fine for debugging, but for production, people will want to integrate their own loggers and not have to specify the logging facility each time.  Not exactly sure what that would look like, but it was definitely out of scope for this PR, so I included it as-is.